sycl/include/sycl/ext/oneapi/matrix/matrix-intel.hpp

```diff
@@ -557,6 +557,348 @@ inline __SYCL_ALWAYS_INLINE void joint_matrix_apply(
 #endif
 }
 
+using namespace sycl::ext::oneapi::experimental::matrix;
+
+// Begin out-of-bounds API
+
+template <typename Group, typename T, size_t NumRows, size_t NumCols, use Use,
+          layout Layout, typename T2>
+inline __SYCL_ALWAYS_INLINE void joint_matrix_fill_checked(
+    Group sg, joint_matrix<Group, T, Use, NumRows, NumCols, Layout> &res,
+    const T2 &v, size_t Stride, size_t Height, size_t Width) {
+#if defined(__SYCL_DEVICE_ONLY__)
+  using storage_element_type =
+      typename oneapi::detail::jm_type_interpretation_helper_trait<
+          T>::storage_element_type;
+  res.spvm = __spirv_CompositeConstructCheckedINTEL<
+      storage_element_type, T, NumRows, NumCols,
+      spv_matrix_use_traits<Use>::value,
+      spv_matrix_layout_traits<Layout>::value>(
+      static_cast<storage_element_type>(v), Stride, Height, Width);
+#else
+  std::ignore = sg;
+  std::ignore = res;
+  std::ignore = v;
+  throw runtime_error("joint matrix is not supported on host device.",
+                      PI_ERROR_INVALID_DEVICE);
+#endif // defined(__SYCL_DEVICE_ONLY__)
+}
+
+template <
+    typename Group, typename S, typename T, size_t NumRows, size_t NumCols,
+    access::address_space Space, access::decorated IsDecorated,
+    std::enable_if_t<std::is_same<S, std::remove_const_t<T>>::value, bool> =
+        true>
+inline __SYCL_ALWAYS_INLINE void
+joint_matrix_load_checked(Group sg,
+                          joint_matrix<Group, S, use::accumulator, NumRows,
+                                       NumCols, layout::dynamic> &res,
+                          multi_ptr<T, Space, IsDecorated> src, size_t stride,
+                          layout Layout, size_t Height, size_t Width) {
+#if defined(__SYCL_DEVICE_ONLY__)
+  static_assert(Space != access::address_space::private_space,
+                "Joint Matrix doesn't support load from private memory!");
+  using DecorT = typename sycl::detail::DecoratedType<T, Space>::type;
+  DecorT *Ptr = sycl::detail::getDecorated<DecorT>(src);
+  switch (Layout) {
+  default:
+    assert(false && "Invalid Memory Layout!");
+  case layout::row_major:
+    res.spvm = __spirv_JointMatrixLoadCheckedINTEL<
+        DecorT, S, NumRows, NumCols,
+        spv_matrix_use_traits<use::accumulator>::value,
+        spv_matrix_layout_traits<layout::dynamic>::value>(
+        Ptr, stride, Height, Width, __spv::MatrixLayout::RowMajor,
```

Comment:
Is it possible to use `spv_matrix_layout_traits<Layout>::value` like below instead of duplicating the code in switch/case?
```suggestion
        Ptr, stride, Height, Width, spv_matrix_layout_traits<Layout>::value,
```

Reply:
We can define something like this then in detail namespace:
```
inline __SYCL_ALWAYS_INLINE __spv::MatrixLayout joint_matrix_layout_to_spv(
    sycl::ext::oneapi::experimental::matrix::layout Layout) {
  switch (Layout) {
  case layout::row_major:
    return __spv::MatrixLayout::RowMajor;
  case layout::col_major:
    return __spv::MatrixLayout::ColumnMajor;
  case layout::ext_intel_packed:
    return __spv::MatrixLayout::Packed;
  case layout::dynamic:
    return __spv::MatrixLayout::Dynamic;
  }
}
```

Reply:
I like this, thank you!

Reply:
@dkhaldi 
> No, it is not possible because layout in this case is a runtime variable.

From https://htmlpreview.github.io/?https://github.com/KhronosGroup/SPIRV-Registry/blob/main/extensions/KHR/SPV_KHR_cooperative_matrix.html :
`MemoryLayout specifies how matrix elements are laid out in memory. It must come from a 32-bit integer constant instruction whose value corresponds to a Cooperative Matrix Layout. See the Cooperative Matrix Layout table for a description of the layouts and detailed layout-specific rules.`

so it contradicts cooperative matrix SPIR-V spec, or am I missing something?

Reply:
I only make the distinction between passing layout in the template and passing it as an argument to the functions. See the two forms in https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/experimental/sycl_ext_matrix/sycl_ext_oneapi_matrix.asciidoc#load

But in both cases, layout must be one of the values:
enum class layout {
  row_major,
  col_major,
  dynamic
};